### ISSUE TITLE
AndroidClientHandler doesn't support proxy servers

### ISSUE THREAD

**Issue by dominik-weber** [xamarin/xamarin-android#2123]
Please add support for HTTP proxy server configuration in AndroidClientHandler :)

**Comment by Cheesebaron** [comment]
This is the related line of code.

https://github.com/xamarin/xamarin-android/blob/3f743678d3e7c19fc6f6e8f6928bfe4ae7d55ff0/src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs#L258

This should probably check whether `Proxy` was set on the HttpClientHandler and use that, then fallback to empty call to `OpenConnection()` if no proxy was set.

**Comment by dominik-weber** [comment]
I know I'm a bit late here, but shouldn't the default behavior (if you just call `new AndroidClientHandler()` without modifying any properties) be to use the system default proxy? Currently these settings are ignored.
@grendello I think support for using the default proxy could be implemented fairly easily by adding something like this to the GetJavaProxy method:
```
            URI java_uri = new URI(EncodeUrl(destination));
            proxy = ProxySelector.Default?.Select(java_uri).FirstOrDefault();
```

**Comment by johnthiriet** [comment]
I totally agree with @dominik-weber and this is what we have implemented in our project in order to be able to debug network calls with Charles Proxy.

**Comment by mzekrallah** [comment]
any update on this guys ?

**Comment by grendello** [comment]
The proxy should be used only if the `Proxy` property is not null (and configured for the specified URL) and the `UseProxy` property is `true` - which is what we implement.

**Comment by dominik-weber** [comment]
@grendello Both `NSUrlSessionHandler` (iOS) and `HttpClientHandler` (.NET) respect the system default proxy, ie. they respect the proxy settings configured in the phone settings app.

See these code examples:

`new HttpClient(new NSUrlSessionHandler()).GetStringAsync("https://google.com");`
-> this will use the proxy settings configured in the iPhone settings app

`new HttpClient(new AndroidMessageHandler()).GetStringAsync("https://google.com");`
-> this will **ignore** the proxy settings configured in the Android settings app

I think the behavior of the different handlers should be consolidated, by default both should use the proxy settings that have been configured system wide.



In the mean time, I'm using this code to manually implement support for the default proxy using the `Java.Net.ProxySelector` API:

```
var handler = new AndroidMessageHandler();

var selector = ProxySelector.Default;
if (selector != null)
    handler.Proxy = new JavaWebProxy(selector);
```

With the following JavaWebProxy class:
```
/// <summary>
/// <see cref="IWebProxy"/> implementation that uses the Java <see cref="ProxySelector"/> API to resolve the proxy information.
/// </summary>
internal class JavaWebProxy : IWebProxy
{
    private readonly ProxySelector _selector;

    public ICredentials Credentials { get; set; }

    public JavaWebProxy(ProxySelector selector)
    {
        _selector = selector ?? throw new ArgumentNullException(nameof(selector));
    }

    public Uri GetProxy(Uri destination)
    {
        var javaUri = new URI(EncodeUrl(destination));

        var proxy = _selector.Select(javaUri).FirstOrDefault();
        if (proxy != null && proxy != Proxy.NoProxy && proxy.Address() is InetSocketAddress address)
        {
            var host = address.HostString;
            var port = address.Port;
            return new UriBuilder("http", host, port).Uri;
        }

        return null;
    }

    public bool IsBypassed(Uri host)
    {
        return GetProxy(host) == null;
    }

    private string EncodeUrl(Uri url)
    {
        // Copied from https://github.com/xamarin/xamarin-android/blob/master/src/Mono.Android/Xamarin.Android.Net/AndroidClientHandler.cs
        // Fixes an issue where urls with unencoded spaces are not recognized by the Java URI class.

        if (url == null)
            return string.Empty;

        // UriBuilder takes care of encoding everything properly
        var bldr = new UriBuilder(url);
        if (url.IsDefaultPort)
            bldr.Port = -1; // Avoids adding :80 or :443 to the host name in the result

        // bldr.Uri.ToString () would ruin the good job UriBuilder did
        return bldr.ToString();
    }
}
```


For reference, this is how you configure the system wide proxy settings on Android:
<img width="401" alt="image" src="https://user-images.githubusercontent.com/1771927/167028886-e303c914-308b-4051-a98c-a235288fd5de.png">

**Comment by dominik-weber** [comment]
UPDATE: Please ignore my last post. This issue has actually been resolved in the past: #3058 #3060. The default proxy settings are used now when `UseProxy == true` and `Proxy == null` (which is the default configuration)
